primers/pcr/pcr.go

```diff
@@ -0,0 +1,184 @@
+
+This package allows users to simulate a PCR reaction or design new primers to
+amplify a given template. This package assumes perfect annealing to template at
+a target temperature, so should only be used for PCR reactions where this is a
+reasonable assumption.
+
+If you are trying to simulate amplification out of a large pool, such as an
+oligo pool, use the `Simulate` rather than `SimulateSimple` function to detect
+if there is concatemerization happening in your multiplex reaction. In most
+other cases, use `SimulateSimple`.
+
+The targetTm in all functions is specifically for Taq polymerase.
+*/
+package pcr
+
+import (
+	"errors"
+	"github.com/TimothyStiles/poly/primers"
+	"github.com/TimothyStiles/poly/transform"
+	"index/suffixarray"
+	"sort"
+	"strings"
+)
+
+// https://doi.org/10.1089/dna.1994.13.75
+var minimalPrimerLength int = 15
+
+// DesignPrimersWithOverhangs designs two primers to amplify a target sequence,
+// adding on an overhang to the forward and reverse strand. This overhang can
+// contain additional DNA needed for assembly, like Gibson assembly overhangs
+// or GoldenGate restriction enzyme sites.
+func DesignPrimersWithOverhangs(sequence, forwardOverhang, reverseOverhang string, targetTm float64) (string, string) {
+	sequence = strings.ToUpper(sequence)
+	forwardPrimer := sequence[0:minimalPrimerLength]
+	for i := 0; primers.MeltingTemp(forwardPrimer) < targetTm; i++ {
+		forwardPrimer = sequence[0 : minimalPrimerLength+i]
+	}
+	reversePrimer := transform.ReverseComplement(sequence[len(sequence)-minimalPrimerLength:])
+	for i := 0; primers.MeltingTemp(reversePrimer) < targetTm; i++ {
+		reversePrimer = transform.ReverseComplement(sequence[len(sequence)-(minimalPrimerLength+i):])
+	}
+
+	// Add overhangs to primer
+	forwardPrimer = forwardOverhang + forwardPrimer
+	reversePrimer = transform.ReverseComplement(reverseOverhang) + reversePrimer
+
+	return forwardPrimer, reversePrimer
+}
+
+// DesignPrimers designs two primers to amplify a target sequence and only that
+// target sequence (no overhangs).
+func DesignPrimers(sequence string, targetTm float64) (string, string) {
+	return DesignPrimersWithOverhangs(sequence, "", "", targetTm)
+}
+
+// SimulateSimple simulates a PCR reaction. It takes in a list of sequences and
+// a list of primers, with support for complex multiplex reactions, produces
+// a list of all possible PCR fragments from such a reaction. It does not
+// detect concatemerization, which could be useful or very detrimental to
+// your reactions. The variable `circular` is for if the target template is
+// circular, like a plasmid.
+func SimulateSimple(sequences []string, targetTm float64, circular bool, primerList []string) []string {
+	// Set all primers to uppercase.
+	for i := range primerList {
+		primerList[i] = strings.ToUpper(primerList[i])
+	}
+
+	var pcrFragments []string
+	for _, sequence := range sequences {
+		sequence = strings.ToUpper(sequence)
```

Comment:
Sanitization of inputs, I think this could be a responsibility of util functions and at the beginning of the SimulateSimple function.

Reply:
Adds additional cognitive overhead, while these functions are quite easy to read / see the sanitation. Would be up to @TimothyStiles , but personally I think having a .ToUpper in a few places isn't that big of a deal.